File: app/src/main/java/za/co/woolworths/financial/services/android/models/network/WfsApiInterceptor.java
package za.co.woolworths.financial.services.android.models.network;

import java.io.IOException;
import java.net.SocketTimeoutException;

import okhttp3.Interceptor;
import okhttp3.MediaType;
import okhttp3.Protocol;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.ResponseBody;
import okio.Buffer;
import za.co.woolworths.financial.services.android.models.dao.ApiRequestDao;
import za.co.woolworths.financial.services.android.models.dao.ApiResponseDao;
import za.co.woolworths.financial.services.android.ui.activities.maintenance.NetworkRuntimeExceptionViewController;
import za.co.woolworths.financial.services.android.util.GZIPCompression;

/**
 * Created by eesajacobs on 2016/12/29.
 */

public class WfsApiInterceptor implements Interceptor {
    public static final String TAG = "WfsApiInterceptor";

    private NetworkConfig networkConfig;

    public WfsApiInterceptor(NetworkConfig networkConfig) {
        this.networkConfig = networkConfig;
    }

    @Override
    public Response intercept(Chain chain) throws IOException {

        Request request = chain.request();
        long t1 = System.nanoTime();

        String requestLog = String.format("Sending request %s on %s%n%s", request.url(), chain.connection(), request.headers());

        String cacheTimeHeaderValue = request.header("cacheTime");
        long cacheTime = Integer.parseInt(cacheTimeHeaderValue == null ? "0" : cacheTimeHeaderValue);//cache time in seconds

        // getForceNetworkUpdate() will force the request to update
        if (cacheTime == 0) {
            try {
                Response originalResponse = chain.proceed(request);
                return originalResponse.newBuilder()
                        .header("Accept-Encoding", "gzip")
                        .build();
            } catch (Exception exception) {
                if (exception instanceof SocketTimeoutException) {
                    new NetworkRuntimeExceptionViewController().openSocketTimeOutDialog();
                }
                throw exception;
            }
        }

        final String endpoint = request.url().toString();
        final String headers = request.headers().toString();
        final String parametersJson = (request.method().compareToIgnoreCase("post") == 0 ? bodyToString(request) : "{}");

        ApiRequestDao apiRequestDao = new ApiRequestDao(cacheTime).get(request.method(), endpoint, headers, parametersJson);
        ApiResponseDao apiResponseDao = new ApiResponseDao().getByApiRequestId(apiRequestDao.id);

        // Override empty request type with current request type
        apiRequestDao.requestType = request.method();

        //cache exists. return cached response
        if (apiResponseDao.id != null && !OneAppService.Companion.getForceNetworkUpdate()) {
            return new Response.Builder()
                    .code(apiResponseDao.code)
                    .message(apiResponseDao.message)
                    .request(chain.request())
                    .protocol(Protocol.HTTP_1_0)
                    .body(ResponseBody.create(MediaType.parse(apiResponseDao.contentType), apiResponseDao.body))
                    .build();
        }

        //cache does not exist. Proceed with service call.
        Response response = null;
        try {
            response = chain.proceed(request);
        } catch (Exception exception) {
            if (exception instanceof SocketTimeoutException) {
                new NetworkRuntimeExceptionViewController().openSocketTimeOutDialog();
            }
            throw exception;
        }

        //save the newly created apiRequestDao
        apiRequestDao.save();

        OneAppService.Companion.setForceNetworkUpdate(false);

        apiResponseDao.apiRequestId = apiRequestDao.id;
        apiResponseDao.message = response.message();
        apiResponseDao.code = response.code();
        apiResponseDao.headers = response.headers().toString();
        apiResponseDao.body = GZIPCompression.decompress(response.body().bytes());

        apiResponseDao.contentType = response.body().contentType().toString();

        //save the newly created apiResponseDao
        apiResponseDao.save();
        long t2 = System.nanoTime();

        String responseLog = String.format("Received response for %s in %.1fms%n%s", apiResponseDao.body + response.request().url(), (t2 - t1) / 1e6d, apiResponseDao.headers);
        return response.newBuilder()
                .header("Cache-Control", "max-age=60")
                .body(ResponseBody.create(MediaType.parse(apiResponseDao.contentType), apiResponseDao.body))
                .build();
    }

    public String bodyToString(final Request request) {
        try {
            final Request copy = request.newBuilder().build();
            final Buffer buffer = new Buffer();
            copy.body().writeTo(buffer);
            return buffer.readUtf8();
        } catch (final IOException e) {
            return "did not work";
        }
    }
}

File: app/src/main/java/za/co/woolworths/financial/services/android/models/network/WfsApiInterceptor.kt
package za.co.woolworths.financial.services.android.models.network

import android.util.Log
import com.awfs.coordination.BuildConfig

import java.io.IOException

import okhttp3.Interceptor
import okhttp3.MediaType
import okhttp3.Protocol
import okhttp3.Request
import okhttp3.Response
import okhttp3.ResponseBody
import okio.Buffer
import za.co.woolworths.financial.services.android.models.WoolworthsApplication
import za.co.woolworths.financial.services.android.models.dao.ApiRequestDao
import za.co.woolworths.financial.services.android.models.dao.ApiResponseDao
import za.co.woolworths.financial.services.android.util.GZIPCompression

/**
 * Created by eesajacobs on 2016/12/29.
 */

class WfsApiInterceptor : Interceptor {

    companion object {
        const val TAG = "WfsApiInterceptor"
    }

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        Log.i(TAG, "inside intercept callback")

        val request = chain.request()

        val t1 = System.nanoTime()

        val requestLog = String.format("Sending request %s on %s%n%s", request.url(), chain.connection(), request.headers())

        val cacheTimeHeaderValue = request.header("cacheTime")
        val cacheTime = Integer.parseInt(cacheTimeHeaderValue
                ?: "0").toLong()//cache time in seconds

        Log.d(TAG, "request\n$requestLog")

        if (cacheTime == 0L) {
            val originalResponse = chain.proceed(request)
            return originalResponse.newBuilder()
                    .header("Accept-Encoding", "gzip")
                    .build()
        }

        val endpoint = request.url().toString()
        var headers = request.headers().toString()

        val parametersJson = if (request.method().compareTo("post", ignoreCase = true) == 0) bodyToString(request) else "{}"

        val apiRequestDao = ApiRequestDao(cacheTime).get(request.method(), endpoint, headers, parametersJson)
        val apiResponseDao = ApiResponseDao().getByApiRequestId(apiRequestDao.id)

        if (apiResponseDao.id != null) {  //cache exists. return cached response
            return Response.Builder()
                    .code(apiResponseDao.code)
                    .message(apiResponseDao.message)
                    .request(chain.request())
                    .protocol(Protocol.HTTP_1_0)
                    .body(ResponseBody.create(MediaType.parse(apiResponseDao.contentType), apiResponseDao.body))
                    .build()
        }

        //cache does not exist. Proceed with service call.
        val response = chain.proceed(request)

        //save the newly created apiRequestDao
        apiRequestDao.save()

        apiResponseDao?.apply {
            apiRequestId = apiRequestDao.id
            message = response.message()
            code = response.code()
            headers = response.headers().toString()
            body = GZIPCompression.decompress(response.body()!!.bytes())

            contentType = response.body()!!.contentType()!!.toString()
            //save the newly created apiResponseDao
            save()

            val t2 = System.nanoTime()
            val responseLog = String.format("Received response for %s in %.1fms%n%s", body + response.request().url(), (t2 - t1) / 1e6, headers)
            Log.d(TAG, "response\n$responseLog")

        }

        return response.newBuilder()
                .header("Cache-Control", "max-age=60")
                .body(ResponseBody.create(MediaType.parse(apiResponseDao.contentType), apiResponseDao.body))
                .build()
    }

    private fun bodyToString(request: Request): String {
        return try {
            val copy = request.newBuilder().build()
            val buffer = Buffer()
            copy.body()!!.writeTo(buffer)
            buffer.readUtf8()
        } catch (e: IOException) {
            "did not work"
        }

    }


}
